### ISSUE TITLE
Adding per-keyspace and percentile Dropwizard metrics

### ISSUE THREAD

**Issue by marksurnin** [stargate/stargate#944]
Currently the `http_server_requests_seconds` Dropwizard metrics that are exposed at `:8084/metrics` look like this:

```
http_server_requests_seconds_count{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",} 172.0
http_server_requests_seconds_sum{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",} 5.934070816
http_server_requests_seconds_max{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",} 0.076524805
```

We'd like to include keyspace name here and record these metrics at common percentiles (p95, p99) so we could set up per-keyspace SLOs. Is this something you'd be open to including as an opt-in config setting?

- - -
Here are some details for reference:

1. To get the keyspace name in the metric, we could substitute it in `uri="/graphql/{keyspaceName}` instead of the template or add it as an extra tag.

I briefly tested the latter approach in [ResourceTagsProvider](https://github.com/stargate/stargate/blob/6452957cd078f17f1c0067c943e3fc2129f61ceb/metrics-jersey/src/main/java/io/stargate/metrics/jersey/ResourceTagsProvider.java#L54) and this tag was added to the metric: `keyspaceName="hello_cassandra"`.

```
private Tag convertPathParamToTag(Map.Entry<String, List<String>> pathParamEntry) {
    String values = String.join(",", pathParamEntry.getValue());
    return Tag.of(pathParamEntry.getKey(), values);
}

@Override
public Iterable<Tag> httpRequestTags(RequestEvent event) {
    ...
    List<Tag> pathParamTags = request.getUriInfo().getPathParameters().entrySet().stream()
        .map(this::convertPathParamToTag)
        .collect(Collectors.toList());

    return defaultTags
        .and(pathParamTags)
        ...
```

1. To record percentile metrics, I added a meter filter to the `meterRegistry` config in [MetricsImpl.java](https://github.com/stargate/stargate/blob/6452957cd078f17f1c0067c943e3fc2129f61ceb/core/src/main/java/io/stargate/core/metrics/impl/MetricsImpl.java#L36). This is just a rough version as a PoC:

```
MeterFilter meterFilter = new MeterFilter() {
      @Override
      public DistributionStatisticConfig configure(Meter.Id id, DistributionStatisticConfig config) {
        if (id.getType() == Meter.Type.TIMER && id.getName().matches("^http.*")) {
          return DistributionStatisticConfig.builder()
              .percentilesHistogram(true)
              .percentiles(0.5, 0.90, 0.95, 0.99)
              .build()
              .merge(config);
        } else {
          return config;
        }
      }
    };
meterRegistry.config().meterFilter(meterFilter);
```

These are the the new metrics
```http_server_requests_seconds{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",quantile="0.5",} 0.043515904
http_server_requests_seconds{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",quantile="0.9",} 0.121110528
http_server_requests_seconds{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",quantile="0.95",} 0.192413696
http_server_requests_seconds{method="POST",module="graphqlapi",status="200",uri="/graphql/{keyspaceName}",quantile="0.99",} 0.225968128

**Comment by ivansenic** [comment]
@marksurnin Percentiles are easy, I will add a configuration option where you can specify it in the form: `0.5,0.9,0.95,0.99`. I might even have `0.99` as the default here.

For the keyspace tags I would also make it configurable, but more in the way that you can specify a path param(s) to include to tags. Then you can add any path related param.

Should be straightforward. Would this be fine for you?

**Comment by marksurnin** [comment]
That sounds great @ivansenic. Thanks a lot!

**Comment by tub** [comment]
@ivansenic Something we've also discussed is having an allow-list of HTTP Headers we can pass in to add extra tags. For example, we have unique client identifiers that would be incredibly useful to graph independently.

(Slightly offtopic, but it came up in the same conversation)

**Comment by ivansenic** [comment]
@tub It's perfectly on topic. We already have a `io.stargate.core.metrics.api.HttpMetricsTagProvider` interface for exactly this purpose already hooked into everything. However the default impl returns empty tags.

The idea is that you can supply your own impl, but we could also go forward here and allow that default impl is based on the config settings where you would pass the header names to grab.

@dimas-b @dougwettlaufer How does this sound?

**Comment by dougwettlaufer** [comment]
> The idea is that you can supply your own impl, but we could also go forward here and allow that default impl is based on the config settings where you would pass the header names to grab.

+1 config driven rather than adding a custom impl sounds good

**Comment by dimas-b** [comment]
Both config + standard tags provide and a customer `HttpMetricsTagProvider` sound good to me.

I wonder, though, if a standard impl. can be flexible enough to cover the majority of use cases. @tub : in your use case, if any transformation of HTTP header values required or a header value corresponds to tag value "as is"?

**Comment by tub** [comment]
We should be able to just use the http header value as-is. I think if anyone needs to do any mangling it's likely they can do it on the metrics collection side, eg in Prometheus relabel configs.

**Comment by ivansenic** [comment]
Percentiles are solved in #974. Expecting system prop in format `stargate.metrics.http_server_requests_percentiles=0.95,0.99,0.999`.

Path and header tags to follow.

**Comment by ivansenic** [comment]
Path parameter extraction done in #1035.  Examples:

- `stargate.metrics.http_server_requests_path_param_tags=*` - all path params
- `stargate.metrics.http_server_requests_path_param_tags=keyspaceName,collection` - only path params that are matching `keyspaceName` or `collection`.

**This is a global option and affects all endpoints.**

Support for headers coming in separate PR, as I need to split this up.
